dali/kernels/common/utils.h

```diff
@@ -16,43 +16,64 @@
 #define DALI_KERNELS_COMMON_UTILS_H_
 
 #include <utility>
-#include "dali/core/host_dev.h"
+#include "dali/core/util.h"
+#include "dali/core/traits.h"
 
 namespace dali {
 namespace kernels {
 
 template <typename T, typename U>
 T Permute(const T& in, const U& permutation) {
   T out = in;
-  for (size_t i = 0; i < permutation.size(); i++) {
+  for (size_t i = 0; i < dali::size(permutation); i++) {
     auto idx = permutation[i];
     out[i] = in[idx];
   }
   return out;
 }
 
-template <typename Shape>
-DALI_HOST_DEV
-void CalcStrides(Shape &strides, const Shape& shape) {
-  strides[strides.size() - 1] = 1;
-  for (int d = strides.size() - 2; d >= 0; d--) {
-    strides[d] = strides[d + 1] * shape[d + 1];
+template <typename Stride, typename Extent>
+inline void CalcStrides(Stride *strides, const Extent *shape, int ndim) {
+  if (ndim > 0) {
+    uint64_t v = 1;
+    for (int i = ndim - 1; i > 0; i--) {
+      strides[i] = v;
+      v *= shape[i];
+    }
+    strides[0] = v;
   }
 }
 
-template <typename OutStrides, typename Shape>
+DALI_NO_EXEC_CHECK
+template <typename T, typename Size>
 DALI_HOST_DEV
-void CalcStrides(OutStrides &out_strides, const Shape& shape) {
-  Shape strides = shape;  // strides has the same size as shape
-  CalcStrides(strides, shape);
-  for (int d = 0; d < static_cast<int>(strides.size()); d++)
-    out_strides[d] = strides[d];
+inline std::enable_if_t<has_resize<T>::value> ResizeIfPossible(T &object, Size size) {
+  object.resize(size);
+}
+
+template <typename T, typename Size>
+DALI_HOST_DEV
+inline std::enable_if_t<!has_resize<T>::value> ResizeIfPossible(T &object, Size size) {}
+
+template <typename Strides, typename Shape>
+DALI_HOST_DEV
+void CalcStrides(Strides &strides, const Shape& shape) {
+  int ndim = dali::size(shape);
+  ResizeIfPossible(strides, ndim);  // no-op if strides is a plain array or std::array
+  if (ndim > 0) {
```

Comment:
Can't you call L36 here?

Reply:
That would assume that I can get a pointer from Strides.